docs: add toolchains and llvm (#398741)
NixOS/nixpkgs: doc/toolchains/llvm.chapter.md

# The LLVM Toolchain {#chap-toolchains}

> LLVM is a target-independent optimizer and code generator and serves as the basis for many languages such as Haskell, Rust, Zig, and many other languages. It forms the base tools for Apple's Darwin platform.

Reviewer: I think it would be more accurate to talk about compilers than languages.

Author: Done

## Using LLVM {#sec-using-llvm}

LLVM has two ways of being used. One is by using it across all of nixpkgs and the other is to compile and build other packages.

### Building packages with LLVM {#sec-building-packages-with-llvm}

> To build a package with LLVM, the simplest way is to invoke it with `pkgsLLVM`. The simplest way is to run:

Reviewer: Could we explain that pkgsLLVM is equivalent to importing Nixpkgs a certain way? I think it would be really good if people knew that `pkgsCross`, `pkgsLLVM` etc. aren't magic.

Author: Yes, I think that'd make sense. I can also show how you can setup LLVM cross compiling via `crossPlatform` when importing nixpkgs.

Author: Done

Reviewer: Could we order it the other way around? Explain what's happening, and then introduce pkgsLLvm as a shorthand?

Author: Ok, it's more verbose as to how things function.

# The LLVM Toolchain {#chap-toolchains}

LLVM is a target-independent optimizer and code generator and serves as the basis for many compilers such as Haskell's GHC, Rust, Zig, and many others. It forms the base tools for Apple's Darwin platform.

## Using LLVM {#sec-using-llvm}

> LLVM has two ways of being used. One is by using it across all of nixpkgs and the other is to compile and build other packages.

Author: Done

### Building packages with LLVM {#sec-building-packages-with-llvm}

> Nixpkgs supports two methods of compiling the world with LLVM, one is via setting `useLLVM` in `crossSystem` while importing. This is the recommended way when doing cross compiling as it is more expressive. An example of doing `aarch64-linux` cross from `x86_64-linux` with LLVM on the target is by doing the following:

Author: Done

```nix
import <nixpkgs> {
  localSystem = {
    system = "x86_64-linux";
  };
  crossSystem = {
    useLLVM = true;
    linker = "lld";
  };
}
```

> Note that we set `linker` to `lld`, this is because LLVM has its own linker called "lld". By setting it, we utilize Clang and lld within this new instance of nixpkgs. The other method which nixpkgs supports for building everything with LLVM is via `pkgsLLVM`. This is recommended to be used with `nix-build` (or `nix build`). An example of this is as simple as the following:

Reviewer: I'm not sure I would describe pkgsLLVM as another method, since it's the same thing. Calling it a shorthand would be better.  I think "easier to use" would be better than "recommended to be used also".

Author: Ok, I've changed it.

> Note that we set `linker` to `lld`, this is because LLVM has its own linker called "lld". By setting it, we utilize Clang and lld within this new instance of nixpkgs. There is a shorthand method which nixpkgs supports for building everything with LLVM is via `pkgsLLVM`. This is easier to be used with `nix-build` (or `nix build`). An example of this is as simple as the following:

Author: Done

```bash
nix-build -A pkgsLLVM.hello
```

> This will compile the GNU hello package with LLVM and the lld linker like previously mentioned. Another simple way is to override the stdenv with `clangStdenv`. This causes a single package to be built with Clang. However, this `stdenv` does not provide compiler-rt, libc++, and libunwind. It will use the platform's default. This is the preferred way to make something use Clang in nixpkgs when packaging a piece of software. There are cases where providing just Clang isn't enough, this is where the `libcxxStdenv` comes in. `libcxxStdenv` uses Clang with libc++ and compiler-rt.

Reviewer: I think it would be good if `useLLVM` and `clangStdenv` were each other their own subheadings.  `clangStdenv` is really easy to miss here because the paragraph it's explained in starts by being about `useLLVM`.

Author: It has been split off into it's own heading now.

# The LLVM Toolchain {#chap-toolchains}

> LLVM is a target-independent optimizer and code generator and serves as the basis for many compilers such as Haskell's GHC, Rust, Zig, and many others. It forms the base tools for Apple's Darwin platform.

Reviewer: ```suggestion
LLVM is a target-independent optimizer and code generator and serves as the basis for many compilers such as Haskell's GHC, rustc, Zig, and many others. It forms the base tools for Apple's Darwin platform.
```

(Does Zig's compiler have a separate name?)

# The LLVM Toolchain {#chap-toolchains}

LLVM is a target-independent optimizer and code generator and serves as the basis for many compilers such as Haskell's GHC, rustc, Zig, and many others. It forms the base tools for Apple's Darwin platform.

## Using LLVM {#sec-using-llvm}

LLVM has two ways of being used. One is by using it across all of nixpkgs and the other is to compile and build individual packages.

### Building packages with LLVM {#sec-building-packages-with-llvm}

Nixpkgs supports two methods of compiling the world with LLVM. One is via setting `useLLVM` in `crossSystem` while importing. This is the recommended way when cross compiling as it is more expressive. An example of doing `aarch64-linux` cross compilation from `x86_64-linux` with LLVM on the target is the following:

```nix
import <nixpkgs> {
  localSystem = {
    system = "x86_64-linux";
  };
  crossSystem = {
    useLLVM = true;
    linker = "lld";
  };
}
```

> Note that we set `linker` to `lld`. This is because LLVM has its own linker called "lld". By setting it, we utilize Clang and lld within this new instance of nixpkgs. There is a shorthand method which nixpkgs supports for building everything with LLVM: `pkgsLLVM`. This is easier to be used with `nix-build` (or `nix build`). This can be as simple as the following:

Reviewer: ```suggestion
Note that we set `linker` to `lld`. This is because LLVM has its own linker called "lld". By setting it, we utilize Clang and lld within this new instance of nixpkgs. There is a shorthand method for building everything with LLVM: `pkgsLLVM`. This is easier to use with `nix-build` (or `nix build`):
```

Author: Done

```bash
nix-build -A pkgsLLVM.hello
```

This will compile the GNU hello package with LLVM and the lld linker like previously mentioned.

#### Using `clangStdenv` {#sec-building-packages-with-llvm-using-clang-stdenv}

> Another simple way is to override the stdenv with `clangStdenv`. This causes a single package to be built with Clang. However, this `stdenv` does not provide compiler-rt, libc++, and libunwind. It will use the platform's default. This is the preferred way to make something use Clang in nixpkgs when packaging a piece of software. There are cases where providing just Clang isn't enough, this is where the `libcxxStdenv` comes in. `libcxxStdenv` uses Clang with libc++ and compiler-rt.

Author: Done
